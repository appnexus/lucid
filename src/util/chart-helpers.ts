import _ from 'lodash';
import * as d3TimeFormat from 'd3-time-format';
import * as d3Time from 'd3-time';

export type Collection = Array<{ [key: string]: any }>;
type Fields = string[] | string;

/**
 * stackByFields
 *
 * D3's `stack` groups each series' data together but we sometimes we want the
 * stacked groups to remain grouped as in the original normalized data. This
 * function helps achieve that.
 *
 * @param {object[]} collection - normalized data you want to operate on
 * @param {string[]} fields - fields to pluck off for the y data
 * @return {array[]} - array of arrays, one for row in the original `collection`
 */
export function stackByFields(
	collection: Collection,
	fields: Fields
): Array<Array<[number, number]>> {
	const fieldsArray = _.castArray(fields);

	return _.map(collection, (d) => {
		return _.reduce<Fields, Array<[number, number]>>(
			fieldsArray,
			(acc, field) => {
				const dataPoint: number = _.get(d, field, 0);

				if (_.isEmpty(acc)) {
					return acc.concat([[0, dataPoint]]);
				}

				const maybeLast = _.last(_.last(acc));
				const last = maybeLast === undefined ? 0 : maybeLast;

				return acc.concat([[last, last + dataPoint]]);
			},
			[]
		);
	});
}

/**
 * extractFields
 *
 * This will return the data in a similar format to stackByFields but without
 * the stacking.
 *
 * @param {object[]} collection - normalized data you want to operate on
 * @param {string[]} fields - fields to pluck off for the y data
 * @return {array[]} - array of arrays, one for each field
 */
export function extractFields(
	collection: Collection,
	fields: Fields,
	minDomainValue: number = 0
): Array<Array<[number, number]>> {
	const fieldsArray = _.castArray(fields);

	return _.map(collection, (d) => {
		return _.map(fieldsArray, (field) => [minDomainValue, _.get(d, field, 0)]);
	});
}

/**
 * groupByFields
 *
 * This will return the data in a similar format to d3Shape.stack
 * but without the stacking of the data.
 *
 * @param {object[]} collection - normalized data you want to operate on
 * @param {string[]} fields - fields to pluck off for the y data
 * @return {array[]} - array of arrays, one for each field
 */
export function groupByFields(collection: Collection, fields: Fields) {
	const fieldsArray = _.castArray(fields);

	return _.map(fieldsArray, (field) => {
		return _.map(collection, field);
	});
}

/**
 * byFields
 *
 * Takes a collection of data and returns an array of all the fields off that
 * collection.
 *
 * @param {object[]} collection
 * @param {string[]} fields
 * @return {array}
 */
export function byFields(collection: Collection, fields: Fields) {
	const fieldArray = _.castArray(fields);

	return _.reduce(
		fieldArray,
		(acc: object[], field) => {
			return acc.concat(_.map(collection, field));
		},
		[]
	);
}

/**
 * nearest
 *
 * Divide and conquer algorithm that helps find the nearest element to `value`
 * in `nums`
 *
 * @param {number[]} nums - sorted array of numbers to search through
 * @param {number} value - value you're trying to locate the nearest array element for
 * @return {number} - the nearest array element to the value
 */
export function nearest(nums: number[], value: number): number | undefined {
	if (nums.length < 2) {
		return _.first(nums);
	}

	if (nums.length === 2) {
		return value > (nums[0] + nums[1]) / 2 ? nums[1] : nums[0];
	}

	const mid = nums.length >>> 1;

	return nums[mid] > value
		? nearest(nums.slice(0, mid + 1), value)
		: nearest(nums.slice(mid), value);
}

/**
 * minByFields
 *
 * Returns the minimum element from a collection by a set of fields.
 *
 * @param {object[]} collection
 * @param {string[]} fields
 * @return {any}
 */
export function minByFields(collection: Collection, fields: Fields) {
	return _.min(byFields(collection, fields));
}

/**
 * maxByFields
 *
 * Returns the maximum element from a collection by a set of fields.
 *
 * @param {object[]} collection
 * @param {string[]} fields
 * @return {any}
 */
export function maxByFields(collection: Collection, fields: Fields) {
	return _.max(byFields(collection, fields));
}

/**
 * maxByFieldsStacked
 *
 * Returns the max sum of a set of fields from a collection
 *
 * @param {object[]} collection
 * @param {string[]} fields
 * @return {any}
 */
export function maxByFieldsStacked(collection: Collection, fields: Fields) {
	const fieldArray = _.castArray(fields);

	const sums = _.reduce(
		collection,
		(acc: number[], item) => {
			return acc.concat(_.sum(_.toArray(_.pick(item, fieldArray))));
		},
		[]
	);

	return _.max(sums);
}

/**
 * discreteTicks
 *
 * Returns `count` evenly spaced, representative values from the `array`.
 *
 * @param {array} array
 * @param {number} size - should be greater than 1
 * @return {array}
 */
export function discreteTicks<T>(
	array: T[],
	count: number | undefined | null = null
) {
	if (!array || _.isNil(count) || array.length <= count) {
		return array;
	}

	const step = (array.length - 1) / Math.max(1, count - 1);

	return _.reduce(
		_.times(count),
		(acc: T[], n) => {
			return acc.concat(array[Math.round(n * step)]);
		},
		[]
	);
}

/**
 * transformFromCenter
 *
 * Scaling paths from their center is tricky. This function
 * helps do that be generating a translate/scale transform
 * string with the correct numbers.
 *
 * @param {number} x - the x data point where you want the path to be centered at
 * @param {number} y - the y data point where you want the path to be centered at
 * @param {number} xCenter - the x coordinate of the center of the path you're trying to transform
 * @param {number} yCenter - the x coordinate of the center of the path you're trying to transform
 * @param {number} scale - number to scale to, 2 would be 2x bigger
 * @return {string} - transform string
 */
export function transformFromCenter(
	x: number,
	y: number,
	xCenter: number,
	yCenter: number,
	scale: number
) {
	return `translate(${(1 - scale) * xCenter + (x - xCenter)}, ${
		(1 - scale) * yCenter + (y - yCenter)
	}) scale(${scale})`;
}

const FORMAT_MILLISECOND = d3TimeFormat.timeFormat('.%L');
const FORMAT_SECOND = d3TimeFormat.timeFormat(':%S');
const FORMAT_MINUTE = d3TimeFormat.timeFormat('%I:%M');
const FORMAT_HOUR = d3TimeFormat.timeFormat('%I %p');
const FORMAT_DAY = d3TimeFormat.timeFormat('%a %d');
const FORMAT_WEEK = d3TimeFormat.timeFormat('%b %d');
const FORMAT_MONTH = d3TimeFormat.timeFormat('%b');
const FORMAT_YEAR = d3TimeFormat.timeFormat('%Y');

/**
 * formatDate
 *
 * This function was written to be used for tick formatting with d3 time
 * scales.
 *
 * @param {date} date - input date
 * @return {string} - formatted date
 */
export function formatDate(date: Date): string {
	return (
		d3Time.timeSecond(date) < date
			? FORMAT_MILLISECOND
			: d3Time.timeMinute(date) < date
			? FORMAT_SECOND
			: d3Time.timeHour(date) < date
			? FORMAT_MINUTE
			: d3Time.timeDay(date) < date
			? FORMAT_HOUR
			: d3Time.timeMonth(date) < date
			? d3Time.timeWeek(date) < date
				? FORMAT_DAY
				: FORMAT_WEEK
			: d3Time.timeYear(date) < date
			? FORMAT_MONTH
			: FORMAT_YEAR
	)(date);
}
