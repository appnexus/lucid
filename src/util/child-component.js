import React from 'react';
import _ from 'lodash';

/**
 * createChildComponent
 *
 * Returns a React Component which does not render anything, but has some useful static functions
 *
 * @return {ReactComponent} which renders `null` but exposes static helper functions
 */
export function createChildComponent (definition={}) {
	return React.createClass({
		propTypes: _.get(definition, 'propTypes', null),
		displayName: _.get(definition, 'displayName', null),
		render: () => null,
		statics: {

			/**
			 * getOwnChildren
			 *
			 * Returns all collective children of elements with this component type
			 *
			 * @param {Array} elements - usually `this.props.children` but can be any list of elements which
			 *	                          might have this type
			 * @return {Array} All the collective children of elements with this component type
			 */
			getOwnChildren(elements) {
				let component = this;
				let children = [];
				React.Children.forEach(elements, (element) => {
					if (element && element.type === component) {
						children = children.concat(React.Children.map(element.props.children, x => x));
					}
				});
				return children;
			},

			/**
			 * getAll
			 *
			 * Returns all elements with this component type
			 *
			 * @param {Array} elements - usually `this.props.children` but can be any list of elements which
			 *	                          might have this type
			 * @return {Array} elements with this component type
			 */
			getAll(elements) {
				let component = this;
				let elementsOfThisType = [];
				React.Children.forEach(elements, (element) => {
					if (element && element.type === component) {
						elementsOfThisType = elementsOfThisType.concat(element);
					}
				});
				return elementsOfThisType;
			},

			/**
			 * getAllProps
			 *
			 * Returns the `props` object of each element with this component type
			 *
			 * @param {Array} elements - usually `this.props.children` but can be any list of elements which
			 *	                          might have this type
			 * @return {Array} `props` objects of elements with this component type
			 */
			getAllProps(elements) {
				let component = this;
				return _.map(component.getAll(elements), 'props');
			},

			getAllAsProps(props, elements) {
				let component = this;
				return _.map(_.flatten([_.get(props, definition.propName, [])]), (childProp) => {
					if (_.isPlainObject(childProp)) {
						return _.assign({}, childProp, {children: childProp.children || childProp.text});
					} else {
						return {
							children: childProp
						};
					}
				}).concat(component.getAllProps(elements));
			}
		}
	});
}
