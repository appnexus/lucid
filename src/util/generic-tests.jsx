import sinon from 'sinon';
import React from 'react';
import { mount, shallow } from 'enzyme';
import assert from 'assert';
import describeWithDOM  from './describe-with-dom';
import _ from 'lodash';
import * as lucid from '../index';

// Common tests for all our components
export function common(Component, {
	getDefaultProps = _.noop,
	selector = null,
	exemptFunctionProps = [],
} = {}) {

	function generateDefaultProps(props={}) {
		return _.assign({}, getDefaultProps(), props);
	}

	describe('[common]', () => {
		it('should have a `displayName` defined', () => {
			assert(Component.displayName);
		});

		it('should pass through styles to the root element', () => {
			let style = {
				backgroundColor: '#f0f',
			};
			const wrapper = shallow(<Component {...generateDefaultProps()} style={style}/>);
			const rootWrapper = selector ? wrapper.find(selector).first() : wrapper.first();
			const rootStyle = rootWrapper.prop('style');
			assert(_.every(style, (val, key) => val === rootStyle[key]), 'root style must contain passed styles');
		});

		it('should pass through `className`', () => {
			const expectedClass = 'rAnDoM';
			const wrapper = shallow(<Component {...generateDefaultProps()} className={expectedClass}/>);
			const rootWrapper = selector ? wrapper.find(selector).first() : wrapper.first();
			const classNames = rootWrapper.prop('className').split(' ');

			assert(_.includes(classNames, expectedClass), `'${classNames}' should include '${expectedClass}'`);
		});

		it('should have an application scoped base class', () => {
			const expectedClass = 'lucid-' + Component.displayName;
			const wrapper = shallow(<Component {...generateDefaultProps()} />);
			const rootWrapper = selector ? wrapper.find(selector).first() : wrapper.first();
			const classNames = rootWrapper.prop('className').split(' ');

			assert(_.includes(classNames, expectedClass), `'${classNames}' should include '${Component.displayName}'`);
		});

		it('should have only application scoped classes', () => {
			const wrapper = shallow(<Component {...generateDefaultProps()} />);
			const rootWrapper = selector ? wrapper.find(selector).first() : wrapper.first();
			const parentClasses = rootWrapper.prop('className').split(' ');
			const childrenClasses = rootWrapper.children().reduce((acc, node) => {
				if (!node.prop('className')) {
					return acc;
				}

				return acc.concat(node.prop('className').split(' '))
			}, []);

			const allClasses = parentClasses.concat(childrenClasses);

			_.forEach(allClasses, className => {
				assert(_.includes(className, `lucid-${Component.displayName}`), `${className} must be scoped`);
			});
		});

		it('should only use onX convention for function proptypes', () => {

			_.forEach(Component.propTypes, (value, key) => {
				// See the following tests from the React source code to figure out how
				// this works: https://github.com/facebook/react/blob/v0.14.7/src/isomorphic/classic/types/__tests__/ReactPropTypes-test.js
				let props = {};
				props[key] = _.noop;

				// Here we test the validation against a noop function, if it doesn't
				// have an error, then it's probably a function. The other possible is
				// that it's an `any`, so we'll test for that below
				const isProbablyFunction = !(value(props, key) instanceof Error);

				// We'll also test out a string to see if that valid
				props[key] = '';

				const isAny = isProbablyFunction && !(value(props, key) instanceof Error);

				// If it's probably a function, and it's not `any`, then we make sure
				// it starts with `on`

				if (isProbablyFunction) {
					assert(isAny || _.startsWith(key, 'on') || _.includes(exemptFunctionProps, key), `${key} must follow onX convention`);
				}

			});

		});

		// Only run this test if it's a public component
		if (!Component._lucidIsPrivate) {
			it('should be available as an exported module from index.js', () => {
				assert(lucid[Component.displayName]);
			});
		}
	});
}

// Common tests for all our icon components
export function icons(Component) {
	describeWithDOM('[icon]', () => {
		it('should pass through isBadge prop to underlying Icon component', () => {
			const wrapper = mount(<Component isBadge={true} />);
			const classNames = wrapper.find('svg').prop('className').split(' ');
			const targetClassName = 'lucid-Icon-is-badge';
			assert(_.includes(classNames, targetClassName), `'${classNames}' should include '${targetClassName}'`);
		});
	});
}

// Common tests for all control components
export function controls(Component, { callbackName, controlSelector , eventType }) {
	// Use DOM tests here since some of our controls use dom events under the hood
	describeWithDOM('[control]', () => {
		it('should callback with `event` and `props`', () => {
			const expectedSpecialProp = 32;
			const props = {
				specialProp: expectedSpecialProp,
				[callbackName]: sinon.spy(),
			};
			const wrapper = mount(<Component {...props} />);

			wrapper.find(controlSelector).simulate(eventType);

			// Last argument should be an object with `uniqueId` and `event`
			const { props: { specialProp }, event } = _.last(props[callbackName].args[0]);

			assert(event, 'missing event');
			assert.equal(specialProp, expectedSpecialProp, 'incorrect or missing specialProp');
		});
	});
}
