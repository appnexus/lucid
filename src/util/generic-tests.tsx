import sinon from 'sinon';
import { parse } from 'path';
import React from 'react';
import PropTypes from 'prop-types';
import { mount, shallow, render } from 'enzyme';
import assert from 'assert';
import _, { each, omit, keys, includes } from 'lodash';
import glob from 'glob';
import { addons, mockChannel } from '@storybook/addons';
import timekeeper from 'timekeeper';

addons.setChannel(mockChannel());

interface ICommonConfig {
	getDefaultProps?: () => {};
	exemptFunctionProps?: string[];
	exemptChildComponents?: string[];
	selectRoot?: any;
	noExport?: boolean;
}

// Common tests for all our components
export function common(Component: any, config: ICommonConfig = {} as any) {
	const {
		getDefaultProps = _.constant({}),
		exemptFunctionProps = [],
		exemptChildComponents = [],
		selectRoot = _.identity,
		noExport = false,
	} = config;

	function generateDefaultProps(props = {}) {
		return _.assign({}, getDefaultProps(), props);
	}

	describe('[common]', () => {
		if (!Component) {
			throw new Error('An undefined component was passed to generic tests.');
		}

		if (Component._isLucidHybridComponent) {
			throw new Error(
				`You're trying to run generic tests on a hybrid component which is bad and won't work and will make you cry. Check your spec files for ${Component.displayName} and import the raw component instead of the hybrid version.`
			);
		}

		it('should have a `displayName` defined', () => {
			assert(Component.displayName);
		});

		it('should pass through styles to the root element', () => {
			const style = {
				backgroundColor: '#f0f',
			};
			const wrapper = shallow(
				<Component {...generateDefaultProps()} style={style} />,
				{ disableLifecycleMethods: true }
			);

			const rootWrapper = selectRoot(wrapper).first();
			const rootStyle = rootWrapper.prop('style');
			assert(
				_.every(style, (val, key) => val === rootStyle[key]),
				'root style must contain passed styles'
			);
		});

		it('should pass through `className`', () => {
			const expectedClass = 'rAnDoM';
			const wrapper = shallow(
				<Component {...generateDefaultProps()} className={expectedClass} />,
				{ disableLifecycleMethods: true }
			);
			const rootWrapper = selectRoot(wrapper).first();
			const classNames = rootWrapper.prop('className').split(' ');

			assert(
				_.includes(classNames, expectedClass),
				`'${classNames}' should include '${expectedClass}'`
			);
		});

		it('should have an application scoped base class', () => {
			const expectedClass = 'lucid-' + Component.displayName;
			const wrapper = shallow(<Component {...generateDefaultProps()} />, {
				disableLifecycleMethods: true,
			});
			const rootWrapper = selectRoot(wrapper).first();
			const classNames = rootWrapper.prop('className').split(' ');

			assert(
				_.includes(classNames, expectedClass),
				`'${classNames}' should include '${Component.displayName}'`
			);
		});

		it('should have only application scoped classes', () => {
			const wrapper = shallow(<Component {...generateDefaultProps()} />, {
				disableLifecycleMethods: true,
			});
			const rootWrapper = selectRoot(wrapper).first();
			const parentClasses = rootWrapper.prop('className').split(' ');
			const childrenClasses = rootWrapper.children().reduce((acc, node) => {
				if (!node.prop('className')) {
					return acc;
				}

				return acc.concat(node.prop('className').split(' '));
			}, []);

			const allClasses = parentClasses.concat(childrenClasses);

			_.forEach(allClasses, (className) => {
				assert(
					_.includes(className, `lucid-${Component.displayName}`),
					`${className} must be scoped`
				);
			});
		});

		describe('function propTypes', () => {
			const funcProps = _.pickBy(
				Component.propTypes,
				(propType) => propType === PropTypes.func
			);

			_.forEach(funcProps, (propType, propName) => {
				it(`${propName} should only use onX convention for function proptypes`, () => {
					assert(
						_.startsWith(propName, 'on') ||
							_.includes(exemptFunctionProps, propName),
						`${propName} must follow onX convention`
					);
				});
			});
		});

		describe('child components', () => {
			// Child components are all function types which start with a capital letter
			const childComponents = _.pickBy(Component, (value, key) => {
				return /^[A-Z]/.test(key) && _.isFunction(value);
			});

			describe('propNames in propTypes', () => {
				_.flow(
					(x) => _.map(x, 'propName'),
					(x) => _.compact(x),
					(x) => _.flatMap(x, _.castArray),
					(x) =>
						_.reject(x, (propName) =>
							_.includes(exemptChildComponents, propName)
						),
					(x) =>
						_.forEach(x, (propName) => {
							it(`should include ${propName} in propTypes`, () => {
								assert(
									Component.propTypes[propName],
									`must include ${propName} in propTypes`
								);
							});
						})
				)(childComponents);
			});
		});

		describe('example testing', () => {
			const fileNames = glob.sync(
				`./src/components/**/${Component.displayName}/*.stories.@(j|t)sx`
			);
			each(fileNames, (path) => {
				const lib = require('../../' + path.replace('.tsx', ''));

				each(omit(lib, ['default']), (Story, name) => {
					it(`should match snapshot(s) for ${name}`, () => {
						let result;
						try {
							result = render(<Story {...Story.args} />, {
								disableLifecycleMethods: true,
							});
							expect(result).toMatchSnapshot();
						} catch (err) {
							expect(err).toMatchSnapshot();
						}
					});
				});
			});

			// Support for older examples
			const exampleFileNames = glob.sync(
				`./src/components/**/${Component.displayName}/examples/*.@(j|t)sx`
			);
			_.each(exampleFileNames, (path) => {
				const lib = require('../../' + path.replace('.tsx', ''));

				const Example = lib.default;

				const title = parse(path).name;
				it(`should match snapshot(s) for ${title}`, () => {
					const shallowExample = shallow(<Example />, {
						disableLifecycleMethods: true,
					});

					// If the root of the example is an instance of the Component under test, snapshot it.
					// Otherwise, look under the root for instances of the Component and snapshot those.
					if (shallowExample.is(Component.displayName)) {
						expect(
							shallow(<Component {...shallowExample.props()} />, {
								disableLifecycleMethods: true,
							})
						).toMatchSnapshot();
					} else {
						shallowExample.find(Component.displayName).forEach((example) => {
							expect(
								shallow(<Component {...example.props()} />, {
									disableLifecycleMethods: true,
								})
							).toMatchSnapshot();
						});
					}
				});
			});
		});
	});
}

// Common tests for all our icon components

interface IIconConfig {
	includeInitialState: boolean;
}

export function icons(Component: any, config: IIconConfig = {} as any) {
	// The default expectation is for every Icon to omit `initailState`,
	// if it is passed through to the underlying element
	const { includeInitialState = false } = config;

	describe('[icon]', () => {
		it('should almost always omit the `initialState` key', () => {
			const wrapper = shallow(<Component initialState={{ testState: true }} />);
			const rootProps = keys(wrapper.first().props());

			expect(includes(rootProps, 'initialState')).toBe(includeInitialState);
		});

		it('should add the correct class for isClickable', () => {
			const wrapper = mount(<Component isClickable={true} />);
			const targetClassName = 'lucid-Icon-is-clickable';
			assert(
				wrapper.find('svg').hasClass(targetClassName),
				`Missing '${targetClassName}' class`
			);
		});
	});
}

// Common tests for all control components
export function controls(
	Component: any,
	{ callbackName, controlSelector, eventType, additionalProps = {} }: any
) {
	// Use DOM tests here since some of our controls use dom events under the hood
	describe('[control]', () => {
		it('should callback with `event` and `props`', () => {
			const expectedSpecialProp = 32;
			const props = {
				specialprop: expectedSpecialProp,
				[callbackName]: sinon.spy(),
				...additionalProps,
			};
			const wrapper = mount(<Component {...props} />);

			wrapper.find(controlSelector).first().simulate(eventType);

			// Last argument should be an object with `uniqueId` and `event`
			const {
				props: { specialprop },
				event,
			}: any = _.last(props[callbackName].args[0]);

			assert(event, 'missing event');
			assert.equal(
				specialprop,
				expectedSpecialProp,
				'incorrect or missing specialProp'
			);
		});
	});
}

// Common tests for all Functional Components
//
// These tests are intended to help us make sure our FCs are shaped corrected.
// They are necessary because there isn't a perfect way to get the defaultProps
// to be factored in correctly yet with React/TypeScript:
// https://github.com/DefinitelyTyped/DefinitelyTyped/issues/30695#issuecomment-474780159
export function functionalComponents(FC: any) {
	// Use DOM tests here since some of our controls use dom events under the hood
	describe('[functionalComponent]', () => {
		it('should have the correct `peek` properties', () => {
			expect(FC.propName === undefined || typeof FC.propName === 'string').toBe(
				true
			);
			expect(
				FC._isPrivate === undefined || typeof FC._isPrivate === 'boolean'
			).toBe(true);

			expect(typeof FC.peek).toBe('object');
			expect(typeof FC.peek.description).toBe('string');
			expect(
				FC.peek.extend === undefined || typeof FC.peek.extend === 'string'
			).toBe(true);
			expect(
				FC.peek.extend === undefined || typeof FC.peek.extend === 'string'
			).toBe(true);
			expect(
				FC.peek.categories === undefined || Array.isArray(FC.peek.categories)
			).toBe(true);
			expect(
				FC.peek.madeFrom === undefined || Array.isArray(FC.peek.madeFrom)
			).toBe(true);
		});
	});
}

export const mockDate = (dateString: string) => {
	timekeeper.freeze(new Date(dateString));
};
