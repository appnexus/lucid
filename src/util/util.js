import _ from 'lodash';
import React from 'react';

export const isDevMode = (function isReactInDev() {
	try {
		React.createClass({});
	} catch(e) {
		if (e.message.indexOf('render') !== 0) {
			return true;  // A nice, specific error message
		}
	}

	return false;  // should never happen, but play it safe.
})();

export const isNode = typeof process === 'object' && process.title === 'node';

export const logger = (function() {
	return checkIsDev() ? {
		log,
		warn,
		error
	} : {
		log: _.noop,
		warn: _.noop,
		error: _.noop
	};
})();

export function checkIsDev() {
	return isDevMode &&
		typeof window !== 'undefined' &&
		!isNode &&
		typeof console !== 'undefined';
}

function log(...args) {

	console.log(...args);

	try {
		// --- Welcome to debugging Lucid ---
		// This error was thrown as a convenience so that you can use this
		// stack to find the callsite that caused this warning to fire.
		throw new Error(args[0]);
	} catch (x) {}

}

function warn(...args) {

	console.warn(...args);

	try {
		// --- Welcome to debugging Lucid ---
		// This error was thrown as a convenience so that you can use this
		// stack to find the callsite that caused this warning to fire.
		throw new Error(args[0]);
	} catch (x) {}

}

function error(...args) {

	console.error(...args);

	try {
		// --- Welcome to debugging Lucid ---
		// This error was thrown as a convenience so that you can use this
		// stack to find the callsite that caused this warning to fire.
		throw new Error(args[0]);
	} catch (x) {}

}
