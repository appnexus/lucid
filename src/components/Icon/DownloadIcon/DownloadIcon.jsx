import _ from 'lodash';
import React from 'react';
import Icon from '../Icon';
import { lucidClassNames } from '../../../util/style-helpers';
import { createClass as createReactClass } from '../../../util/component-types';

const cx = lucidClassNames.bind('&-DownloadIcon');

/**
 *
 * {"categories": ["visual design", "icons"], "extend": "Icon", "madeFrom": ["Icon"]}
 *
 * Typically used to denote that something is available for download.
 */
const DownloadIcon = createReactClass({
	displayName: 'DownloadIcon',
	propTypes: {
		...Icon.propTypes,
	},

	render() {
		const { className, ...passThroughs } = this.props;

		return (
			<Icon
				{...passThroughs}
				{..._.pick(passThroughs, _.keys(Icon.propTypes))}
				className={cx('&', className)}
			>
				<path d="M16,12v2.75c0,0.69-0.56,1.25-1.25,1.25H1.25C0.56,16,0,15.44,0,14.75V12h1v1.75C1,13.888,1.112,14,1.25,14h13.5 c0.138,0,0.25-0.112,0.25-0.25V12H16z M8,11c0,0,0.717,0,1.246-0.623C9.249,10.379,13,6,13,6l-1.288-1L9,8.271V1 c0-0.552-0.447-1-1-1S7,0.448,7,1v7.271L4.288,5L3,6c0,0,3.751,4.379,3.754,4.377C7.283,11,8,11,8,11z" />
			</Icon>
		);
	},
});

export default DownloadIcon;
