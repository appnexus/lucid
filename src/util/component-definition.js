import React from 'react';
import _ from 'lodash';
import { createChildComponent } from './child-component';

export function createLucidComponentDefinition(componentDefinition) {
	let {
		arrayOf,
		shape,
		oneOfType,
		node
	} = React.PropTypes;
	let reducers = _.get(componentDefinition, 'reducers');
	let childProps = _.get(componentDefinition, 'childProps', {});
	let childPropsList = _.reduce(childProps, (list, value, key) => list.concat({
		propTypes: value,
		baseName: key,
		propName: key,
		displayName: _.flatten([_.get(componentDefinition, 'displayName', [])]).concat([key]).join('.'),
	}), []);

	return _.assign({}, _.omit(componentDefinition, ['reducers', 'childProps']), {
		propTypes: _.assign(
			{},
			_.get(componentDefinition, 'propTypes'),
			_.reduce(
				childPropsList,
				(newPropTypes, childProp) => _.assign(
					{},
					newPropTypes,
					{[childProp.propName]: oneOfType([arrayOf(shape(childProp.propTypes)), shape(childProp.propTypes), node])}
				),
				{}
			)
		),
		statics: _.assign(
			{},
			_.get(componentDefinition, 'statics'),
			(reducers ? {reducers} : null),
			_.reduce(
				childPropsList,
				(exportedChildComponents, childProp) => _.assign(
					{},
					exportedChildComponents,
					{[childProp.baseName]: createChildComponent(childProp)}
				),
				{}
			)
		)
	});
}
