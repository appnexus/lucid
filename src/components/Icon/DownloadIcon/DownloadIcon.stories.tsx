import React from 'react';
import { Story, Meta } from '@storybook/react';
import { IIconProps } from '../Icon';
import { DownloadIcon } from './DownloadIcon';

export default {
	title: 'Icons/Icons/DownloadIcon',
	component: DownloadIcon,
} as Meta;

//👇 We create a “template” of how args map to rendering
const Template: Story<IIconProps> = (args) => <DownloadIcon {...args} />;

//👇 Each story then reuses that template
export const Primary = Template.bind({});
