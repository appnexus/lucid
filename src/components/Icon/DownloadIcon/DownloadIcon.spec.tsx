import { icons, common } from '../../../util/generic-tests';

import DownloadIcon from './DownloadIcon';

describe('DownloadIcon', () => {
	common(DownloadIcon);
	icons(DownloadIcon);
});
