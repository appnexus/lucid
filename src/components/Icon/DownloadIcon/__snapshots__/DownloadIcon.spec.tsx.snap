// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DownloadIcon [common] example testing should match snapshot(s) for 1.default 1`] = `
<Icon
  aspectRatio="xMidYMid meet"
  className="lucid-DownloadIcon"
  color="primary"
  isClickable={false}
  isDisabled={false}
  onClick={[Function]}
  onSelect={[Function]}
  size={16}
  viewBox="0 0 16 16"
>
  <path
    d="M8 0v12m5-5l-5 5-5-5"
  />
  <path
    d="M.5 13.5v2h15v-2"
  />
</Icon>
`;
